File: Trabajo_Practico_2024/python/frecuencias_palabras/FrecuenciasPalabras.py
import re  
from collections import Counter  # Importar Counter para contar elementos fácilmente
import time  # Importar time para medir el tiempo de ejecución

def contar_palabras(nombre_archivo):
    # Definir un conjunto de palabras que se van a filtrar (conectores y pronombres)
    conectores_pronombres = {
        'y', 'e', 'ni', 'que', 'o', 'u', 'pero', 'mas', 'sino', 'aunque', 'tal', 'es',
        'porque', 'pues', 'como', 'si', 'cuando', 'mientras', 'tan', 'tanto', 'más',
        'según', 'donde', 'a', 'ante', 'bajo', 'cabe', 'con', 'contra', 'de', 'al',
        'desde', 'en', 'entre', 'hacia', 'hasta', 'para', 'por', 'según', 'un', 'qué',
        'sin', 'so', 'sobre', 'tras', 'yo', 'tú', 'él', 'ella', 'nosotros', 'del',
        'nosotras', 'vosotros', 'vosotras', 'ellos', 'ellas', 'me', 'te', 'su',
        'se', 'nos', 'os', 'mío', 'mía', 'míos', 'mías', 'tuyo', 'tuya', 'una',
        'tuyos', 'tuyas', 'suyo', 'suya', 'suyos', 'suyas', 'nuestro', 'mis', 'mas',
        'nuestra', 'nuestros', 'nuestras', 'vuestro', 'vuestra', 'vuestros', 'mi',
        'vuestras', 'este', 'esta', 'estos', 'estas', 'ese', 'esa', 'esos', 'los',
        'esas', 'aquel', 'aquella', 'aquellos', 'aquellas', 'esto', 'eso', 'la', 'el',
        'aquello', 'lo', 'le', 'les'
    }
    
    # Abrir el archivo y leer su contenido
    with open(nombre_archivo, 'r', encoding='utf-8') as archivo:
        texto = archivo.read()
    
    # Encontrar todas las palabras en el texto usando una expresión regular para limpiar las basuras, texto/inicio/letrasodigitos/fin
    palabras = re.findall(r'\b\w+\b', texto.lower())
    
    # Filtrar las palabras que no están en el conjunto de conectores y pronombres
    palabras_filtradas = [palabra for palabra in palabras if palabra not in conectores_pronombres]
    
    # Contar la frecuencia de cada palabra filtrada 
    conteo_palabras = Counter(palabras_filtradas)
    
    return conteo_palabras

def mostrar_palabras_frecuentes(conteo_palabras, n=10):
    # Mostrar las palabras más frecuentes utilizando 
    print(f"Las {n} palabras más frecuentes son:")
    for palabra, conteo in conteo_palabras.most_common(n):
        print(f"{palabra}: {conteo}")

if __name__ == "__main__":
    nombre_archivo = input("Ingresa el nombre del archivo: ")  # Pedir al usuario el nombre del archivo
    
    start_time = time.time()  # Iniciar el conteo de tiempo de ejecución
    
    conteo = contar_palabras(nombre_archivo)  # Contar las palabras en el archivo
    mostrar_palabras_frecuentes(conteo)  # Mostrar las palabras más frecuentes
    
    end_time = time.time()  # Finalizar el conteo de tiempo de ejecución
    calc_time = end_time - start_time  # Calcular el tiempo transcurrido
    print(f"Tiempo de ejecución: {calc_time} segundos")  # Mostrar el tiempo de ejecución total

File: Trabajo_Practico_2024/ruby/frecuencias_palabras/FrecuenciasPalabras.rb
require 'benchmark'

# Definir el conjunto de conectores y pronombres a excluir
CONECTORES_PRONOMBRES = [
  'y', 'e', 'ni', 'que', 'o', 'u', 'pero', 'mas', 'sino', 'aunque', 'tal', 'es', 's', 'n',
  'porque', 'pues', 'como', 'si', 'cuando', 'mientras', 'tan', 'tanto', 'más', 'est',
  'según', 'donde', 'a', 'ante', 'bajo', 'cabe', 'con', 'contra', 'de', 'al', 'qu',
  'desde', 'en', 'entre', 'hacia', 'hasta', 'para', 'por', 'según', 'un', 'qué',
  'sin', 'so', 'sobre', 'tras', 'yo', 'tú', 'él', 'ella', 'nosotros', 'del',
  'nosotras', 'vosotros', 'vosotras', 'ellos', 'ellas', 'me', 'te', 'su', 'm',
  'se', 'nos', 'os', 'mío', 'mía', 'míos', 'mías', 'tuyo', 'tuya', 'una', 's',
  'tuyos', 'tuyas', 'suyo', 'suya', 'suyos', 'suyas', 'nuestro', 'mis', 'mas',
  'nuestra', 'nuestros', 'nuestras', 'vuestro', 'vuestra', 'vuestros', 'mi',
  'vuestras', 'este', 'esta', 'estos', 'estas', 'ese', 'esa', 'esos', 'los',
  'esas', 'aquel', 'aquella', 'aquellos', 'aquellas', 'esto', 'eso', 'la', 'el',
  'aquello', 'lo', 'le', 'les'
]

def contar_palabras(nombre_archivo, conectores_pronombres)
  frecuencia_palabras = {}

  # Verificar si el archivo existe
  if File.exist?(nombre_archivo)
    # Abrir el archivo en modo lectura
    File.open(nombre_archivo, "r") do |archivo|
      # Iterar sobre cada línea del archivo
      archivo.each_line do |linea|
        # Dividir la línea en palabras y convertirlas a minúsculas
        palabras = linea.downcase.scan(/\w+/)
        
        # Filtrar palabras que no están en conectores_pronombres
        palabras_filtradas = palabras.reject { |palabra| conectores_pronombres.include?(palabra) }

        # Contar la frecuencia de cada palabra filtrada
        palabras_filtradas.each do |palabra|
          frecuencia_palabras[palabra] ||= 0
          frecuencia_palabras[palabra] += 1
        end
      end
    end

    # Ordenar las palabras por frecuencia en orden descendente
    palabras_ordenadas = frecuencia_palabras.sort_by { |palabra, conteo| -conteo }

    # Mostrar las 10 palabras más frecuentes
    puts "Las 10 palabras más frecuentes en '#{nombre_archivo}' son:"
    palabras_ordenadas[0..9].each do |palabra, conteo|
      puts "- #{palabra}: #{conteo}"
    end

  else
    puts "El archivo '#{nombre_archivo}' no existe."
  end
end

if __FILE__ == $PROGRAM_NAME
  # Pedir al usuario que ingrese el nombre del archivo
  puts "Ingresa el nombre del archivo: "
  nombre_archivo = gets.chomp

  # Medir tiempo de ejecución
  tiempo_ejecucion = Benchmark.realtime do
    contar_palabras(nombre_archivo, CONECTORES_PRONOMBRES)
  end

  puts "Tiempo de ejecución: #{tiempo_ejecucion.round(8)} segundos"
end
